### PR TITLE
Compressed Sparse Row (CSR) representation for Hamiltonians and Tensors

### PR DESCRIPTION
**Pull request by AmintorDusko**
**Context:**
This PR changes the Pennylane sparse representation to Compressed Sparse Row (CSR) representation.

**Description of the Change:**
When converting a Hamiltonian to a sparse format, this format will be CSR.
The COO format previously used in all parts of PennyLane was left as an internal format in parts of the code where it will be beneficial.

**Benefits:**
The expval calculation speed of Sparse Hamiltonians has increased by 20%-30%.

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by AmintorDusko** [comment]
I didn't review the tests yet. There will be fails.

**Comment by AmintorDusko** [comment]
Some benchmarks below:
dense
sparse

The sparse Hamiltonian representation show a increase in speed of ~25% in the workloads considered.
No change in time was observed for the dense Hamiltonian case, as expected, since the fixed costs are very similar in both formats.

**Comment by soranjh** [comment]
> Some benchmarks below: dense sparse
> 
> The sparse Hamiltonian representation shows an increase of ~25% in speed with the CSR representation. As expected, we observe just a slight improvement in time for dense Hamiltonian cases.

Thanks @AmintorDusko. I am curious how CRS compares with COO in terms of the **time** needed to create the sparse Hamiltonian matrix and also the **size** of the sparse matrix.

**Comment by AmintorDusko** [comment]
> > The sparse Hamiltonian representation shows an increase of ~25% in speed with the CSR representation. As expected, we observe just a slight improvement in time for dense Hamiltonian cases.
> 
> Thanks @AmintorDusko. I am curious how CRS compares with COO in terms of the **time** needed to create the sparse Hamiltonian matrix and also the **size** of the sparse matrix.

Hi @soranjh, thanks for asking.
The time to create the sparse Hamiltonians is essentially the same. Mainly because I kept the COO representation internally to build the Hamiltonian and convert to CSR after the Hamiltonian is built. Scipy conversion from COO to CSR is very efficient.
Regarding the size, CSR, due to its internal structure, will usually be smaller or in the worst case have the COO size.
I'm attaching some images to illustrate what I'm saying. 
time
size

File: doc/releases/changelog-dev.md
Comment (josh146) [inline doc/releases/changelog-dev.md:77]: @AmintorDusko could you briefly expand on why this was done, and the impact :)

Comment (AmintorDusko) [reply doc/releases/changelog-dev.md]: Sure! Thank you for your suggestion, @josh146.

**Review by antalszava — APPROVED** [review]
Looks good :tada: The changes look good, though it would be great to get an extra pair of eyes here before merging.

**Comment by AmintorDusko** [comment]
> Looks good tada The changes look good, though it would be great to get an extra pair of eyes here before merging.

Thank you, @antalszava! I will wait for one more approval before merging.

File: pennylane/devices/default_qubit.py
Comment (soranjh) [inline pennylane/devices/default_qubit.py:507]: Why the `coo` format is still used here? Is it because `csr` is used only for the expval computation of the full sparse Hamiltonian matrix?

File: pennylane/operation.py
Comment (soranjh) [inline pennylane/operation.py:2004]: Can it be `format="csr"`?

**Review by soranjh — APPROVED** [review]
Looks good to me @AmintorDusko, just left two questions.

File: pennylane/operation.py
Comment (AmintorDusko) [reply pennylane/operation.py:2004]: In my tests, I found a very similar time between the two approaches.
I'm happy with changing this, though.

**Comment by AmintorDusko** [comment]
Thank you, @soranjh, @antalszava and @josh146!

**Pull request merged**
